### PR TITLE
Feat: upgrade base image python and cuda versions

### PR DESCRIPTION
**Pull request by fujikosu**
# Purpose

This PR upgrades base image's python version and cuda version in the Dockerfiles for the `notebooks`, `src/sample_cpu_project`, and `src/sample_pytorch_gpu_project` directories.

python:3.9.19 -> python:3.11.10
nvidia/cuda:12.0.1-cudnn8-runtime-ubuntu20.04 -> nvidia/cuda:12.6.1-cudnn-runtime-ubuntu24.04

python 3.11 is the latest version that has Maintenance status==security and python 3.9 will be end of support about in a year and that motivated this update.

updated devcontainers were tested locally both on CPU and GPU machines.

Besides, roadmap section was updated as common requirements-dev.txt for code quality dependencies was already in place and tensorflow GPU example is not a priority anymore

## Does this introduce a breaking change?

* [ ] Yes
* [x] No

## Author pre-publish checklist

* [x] No PII in logs or output
* [x] Made corresponding changes to the documentation
* [x] All new packages used are included in requirements.txt
* [x] Functions use type hints, and there are no type hint errors

## Pull Request Type

What kind of change does this Pull Request introduce?

* [ ] Bugfix
* [x] Feature
* [ ] Code style update (formatting, local variables)
* [ ] Refactoring (no functional changes, no api changes)
* [ ] Documentation content changes
* [ ] Experiment notebook

### CODE REVIEW

File: notebooks/.devcontainer/Dockerfile [hunk 1]
```diff
@@ -1,7 +1,7 @@
-FROM python:3.9.19
+FROM python:3.11.10
 # create non-root user and set the default user
 ARG USERNAME=devuser
-ARG USER_UID=1000
+ARG USER_UID=1010
```
Comment (bhavikm) [inline notebooks/.devcontainer/Dockerfile:4]: just for my understanding, what is the need to update this UID?

File: notebooks/.devcontainer/Dockerfile [hunk 2]
```diff
@@ -15,7 +15,7 @@ USER $USERNAME
 
 # make all python tools installed by pip accesible
 ENV PATH=$PATH:/home/$USERNAME/.local/bin
-RUN pip install pip --upgrade
+RUN pip install --no-cache-dir pip --upgrade
```
Comment (bhavikm) [inline notebooks/.devcontainer/Dockerfile:18]: just for my understanding, why do we need --no-cache-dir?

**Review by bhavikm — APPROVED** [review]

File: notebooks/.devcontainer/Dockerfile (hunk 2 above)
Comment (fujikosu) [reply notebooks/.devcontainer/Dockerfile:18]: To reduce the final container image size. check this link out for more details https://stackoverflow.com/questions/45594707/what-is-pips-no-cache-dir-good-for

File: notebooks/.devcontainer/Dockerfile (hunk 1 above)
Comment (fujikosu) [reply notebooks/.devcontainer/Dockerfile:4]: gpu side's uuid was updated to 1010 due to the reason commented in dockerfile. To keep the file permission to the same user across two containers, I needed to update uuid on this container to the same id with gpu's one

```
# uid=1000 is already used for ubuntu so use 1010 instead
ARG USER_UID=1010
```

**Pull request merged**